### PR TITLE
Created CoSENTLoss.py

### LINKED ISSUE #2438: Implement CoSENT loss
Hello!

## Context
This issue describes a feature that I am planning to be included in a release before v3, or alternatively, in v3 of Sentence Transformers.

## Details
Some [recent work by Li & Li, 2023](https://arxiv.org/abs/2309.12871) brought my attention to CoSENT ([blogpost](https://kexue.fm/archives/8847), [GitHub repo](https://github.com/bojone/CoSENT)). It proposes a new loss function with the purpose of more effectively and directly causing the cosine similarity of a positive pair to be larger than that of a negative pair.

Implementation snippets: 
* in Keras: https://github.com/bojone/CoSENT/blob/124c368efc8a4b179469be99cb6e62e1f2949d39/cosent.py#L79-L90
* in Torch: https://github.com/SeanLee97/AnglE/blob/ce6d402c794fe3ddd1c1c08a756124cd04f6bb65/angle_emb/angle.py#L67-L77

@johneckberg has expressed an interest in contributing this loss.

cc @bwanglzu @ir2718 @johneckberg @aamir-s18 as I know you're interested in my TODO list. John will lead this development, and I will review his PR.

- Tom Aarsen

### PR DESCRIPTION
**Pull request by johneckberg**
# PR Overview

- Resolve #2438

# Details

- Implemented CoSENT loss as described in this [blogpost](https://kexue.fm/archives/8847) and the [AnglE repo](https://github.com/SeanLee97/AnglE/blob/ce6d402c794fe3ddd1c1c08a756124cd04f6bb65/angle_emb/angle.py#L67-L77).
- This loss follows the same input format as CosineSimilarityLoss, requiring each input to contain a pair of sentences and a floating point label, allowing for applications beyond two-category data. To use this loss with MNRLoss, like done [here in the AnglE repo,](https://github.com/SeanLee97/AnglE/blob/ce6d402c794fe3ddd1c1c08a756124cd04f6bb65/angle_emb/angle.py#L489-505) an [internal method](https://github.com/SeanLee97/AnglE/blob/ce6d402c794fe3ddd1c1c08a756124cd04f6bb65/angle_emb/angle.py#L120-133) must be used. I can't think of a way to use a different input format without assumptions about the labels, which would limit the applications beyond two-category data. If anyone has any ideas, take a look at the code and leave a comment!
- Subtraction by 1e^12 is how irrelevant values in the scores matrix become negligible in the logsumexp. I believe this is the constant used by the original implementation, and no other implementations have seen a need to change it. I agree, and for consistency have also used 1e^12. 
- The example in the doc string uses InputExamples, not a Dataset/DatasetDict, which will need to change in the future.

### CODE REVIEW

**Comment by tomaarsen** [comment]
Hello!

I've ran various scripts that use CosineSimilarityLoss with both that loss & CoSENT, and the Spearman correlation coefficient for cosine similarity against the STSBenchmark test set almost universally shows improvements for CoSENT versus CosineSimilarityLoss.
This is very exciting! I'll be diving into the exact implementation details & blogpost soon.

* [train_sts_indomain_semantic.py](https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/data_augmentation/train_sts_indomain_semantic.py)
  * Cosine: 84.21 Spearman correlation based on cosine similarity
  * CoSENT: **85.47** Spearman correlation based on cosine similarity
* [training_multi-task.py](https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/other/training_multi-task.py)
  * Cosine: 81.51 Spearman correlation based on cosine similarity
  * CoSENT: **82.41** Spearman correlation based on cosine similarity
* [training_stsbenchmark_continue_training.py](https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/sts/training_stsbenchmark_continue_training.py)
  * Cosine: 86.35 Spearman correlation based on cosine similarity
  * CoSENT: **87.01** Spearman correlation based on cosine similarity
* [training_stsbenchmark.py](https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/sts/training_stsbenchmark.py)
  * Cosine: 79.18 Spearman correlation based on cosine similarity
  * CoSENT: **84.25** Spearman correlation based on cosine similarity

- Tom Aarsen

**Comment by tomaarsen** [comment]
I think this might already be ready for merging! What do you think @johneckberg?

**Comment by johneckberg** [comment]
@tomaarsen Go ahead! If you think that the documentation is clear and easy enough to follow, then I also think it's ready for merging.

File: sentence_transformers/losses/CoSENTLoss.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import torch
+from torch import nn, Tensor
+from typing import Iterable, Dict
+from ..SentenceTransformer import SentenceTransformer
+from .. import util
+
+
+class CoSENTLoss(nn.Module):
+    """
+    This class implements CoSENT (Consistent Sentence Embedding via Similarity Ranking) loss.
```
Comment (tomaarsen) [inline sentence_transformers/losses/CoSENTLoss.py:10]: Is CoSENT short for Consistent Sentence Embedding via Similarity Ranking? I thought the blogpost/paper just mentioned "Cosine Sentence".

**Review by tomaarsen — COMMENTED** [review]
I've added CoSENTLoss to the docs & updated the formatting slightly. I also mentioned my experiments where CoSENTLoss outperformed CosineSimilarityLoss.

---

cc: @bojone You might be interested in this work; we're impressed with the performance of CoSENT, and would like to implement it into Sentence Transformers to improve the accessibility of your loss function.

- Tom Aarsen

File: sentence_transformers/losses/CoSENTLoss.py (hunk 1 above)
Comment (johneckberg) [reply sentence_transformers/losses/CoSENTLoss.py:10]: I think you are correct, I see now where the author calls it Cosine Sentence. I was going off the name in [this](https://github.com/RingBDStack/CoSENT/blob/main/README.md) repo, but on second inspection I don't see any source for their choice of name.

Comment (johneckberg) [reply sentence_transformers/losses/CoSENTLoss.py]: Just fixed the doc string!

**Comment by bojone** [comment]
@tomaarsen Wonderfull! I'm very happy to see CoSENT receiving further recognition.
